iml-wasm-components/iml-action-dropdown/src/lib.rs

```diff
@@ -0,0 +1,146 @@
+#[macro_use]
+extern crate seed;
+use seed::prelude::*;
+use std::collections::{HashMap, HashSet};
+use wasm_bindgen::JsCast;
+use web_sys::console::log_1;
+use web_sys::Node;
+use serde;
+
+type Record = (i64, i64);
+
+pub type Locks = HashMap<String, HashSet<LockChange>>;
+
+/// The type of lock
+#[derive(serde::Deserialize, serde::Serialize, Debug, Eq, PartialEq, Hash, Clone)]
+#[serde(rename_all = "lowercase")]
+enum LockType {
+    Read,
+    Write,
+}
+
+/// The Action associated with a `LockChange`
+#[derive(serde::Deserialize, serde::Serialize, Debug, Eq, PartialEq, Hash, Clone)]
+#[serde(rename_all = "lowercase")]
+enum Action {
+    Add,
+    Remove,
+}
+
+/// A change to be applied to `Locks`
+#[derive(serde::Deserialize, serde::Serialize, Debug, Eq, PartialEq, Hash, Clone)]
+pub struct LockChange {
+    job_id: u64,
+    content_type_id: u64,
+    item_id: u64,
+    description: String,
+    lock_type: LockType,
+    action: Action,
+}
+
+// Model
+
+#[derive(Clone)]
+struct Model {
+    records: Vec<Record>,
+    locks: Locks,
+    open: bool,
+}
+
+// Setup a default here, for initialization later.
+impl Default for Model {
+    fn default() -> Self {
+        Self {
+            records: vec![],
+            locks: HashMap::new(),
+            open: false,
+        }
+    }
+}
+
+// Update
+
+#[derive(Clone)]
+enum Msg {
+    Open(bool),
+    Locks(Locks)
+}
+
+/// The sole source of updating the model; returns a fresh one.
+fn update(msg: Msg, model: Model) -> Update<Model> {
+    match msg {
+        Msg::Open(open) => Render(Model { open, ..model }),
+        Msg::Locks(locks) => Render(Model { locks, ..model })
+    }
+}
+
+// View
+fn show_class(model: &Model) -> &str {
+    if model.open {
+        "show"
+    } else {
+        ""
+    }
+}
+
+/// The top-level component we pass to the virtual dom.
+fn view(state: seed::App<Msg, Model>, model: &Model) -> El<Msg> {
+    // Attrs, Style, Events, and children may be defined separately.
+    let outer_style = style!{
+            "display" => "flex";
+            "flex-direction" => "column";
+            "text-align" => "center"
+    };
+
+    let next_open = Msg::Open(!model.open);
+
+    let recordEls: Vec<El<Msg>> = model.records.clone().into_iter()
+        .map(|x| div![
+                    format!("{:?}", x)
+                ])
+                .collect();
+
+    div![
+        outer_style,
+            recordEls,
+        div![
+            class!["btn-group", show_class(&model)],
+            button![
+                class!["btn", "btn-primary", "dropdown-toggle"],
+                mouse_ev("click", move |ev| {
+                    ev.stop_propagation();
+                    ev.prevent_default();
+                    next_open.clone()
+                }),
+                "Actions"
+            ],
+            div![
```

Comment:
Is this where we will put the record elements?

Reply:
This is where the actions will go.